Refactored contributors guide

awanlin opened the pull request:
## Hey, I just made a Pull Request!

This PR attempts to refactor the contributors guide based on feedback on Discord: https://discord.com/channels/687207715902193673/1145821027314716742

The following was changed:

- removed mention of cloning the repo directly as we want everyone to work from a fork and included a link to the GitHub docs on this subject
- added the Coding Guidelines, Creating Changesets, and DCS sections from: https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md
- added a new sections about API Reports, Submitting a Pull Request, and the Review Process

This does introduce some duplication and I would like to change the `CONTRIBUTING.md` file to act as more a pointer to the documentation in `contributors.md` but I'd like some feedback first. My opinion is that the content on Backstage.io is where most people start or are generally looking for documentation so having this there seems to be the best place for it.

I'm also wondering if this should be it's own top level section with separate sections for the various topics?

#### :heavy_check_mark: Checklist

<!--- Please include the following in your Pull Request when applicable: -->

- [ ] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [x] Added or updated documentation
- [ ] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

awanlin commented:
Hi @adamdmharvey, if you happen to have time your feedback on this one would be very welcome 👍

adamdmharvey commented on docs/getting-started/contributors.md:
```diff
@@ -78,7 +75,19 @@ serves the react app ([0]) and the backend ([1]);
+
+Also be sure to skim through our [ADRs](docs/architecture-decisions) to see if they cover what you're working on. In particular [ADR006: Avoid React.FC and React.SFC](docs/architecture-decisions/adr006-avoid-react-fc.md) is one to look out for.
+
+If there are any updates in `markdown` file please make sure to run `yarn run lint:docs`. Though it is checked on `lint-staged`. It is required to install [vale](https://docs.errata.ai/vale/install) separately and make sure it is accessed by global command.
```
```suggestion
If there are any updates in any Markdown files please make sure to run `yarn run lint:docs`. (Though it is also checked on `lint-staged` via the CI/CD pipeline.) It is required to install [Vale](https://docs.errata.ai/vale/install) separately and make sure it is accessed by global command.
```

awanlin replied:
Ah so this isn't totally correct, `lint-staged` runs once just before the changes get committed to your branch, I've adjusted the wording to fit that so maybe take another look to make sure it seems reasonable 👍

adamdmharvey commented on docs/getting-started/contributors.md:
```diff
@@ -142,7 +152,29 @@ default app configs.
+
+In general, changesets are only needed for changes to packages within `packages/` or `plugins/` directories, and only for the packages that are not marked as `private`. Changesets are also not needed for changes that do not affect the published version of each package, for example changes to tests or in-line source code comments.
+
+Changesets **are** needed for new packages, as that is what triggers the package to be part of the next release.
```
i feel like somewhere in here we should note if you update the README we should get a changeset. i myself have questioned that before, and i think the prevailing feedback is usually we want any README update published to Npm's web site and thus a changeset is preferred?

awanlin replied:
Yeah, I agree but it didn't get in as I'm not sure how to word it within what already exists, if you have some suggestions I'm happy to add them!

adamdmharvey commented on docs/getting-started/contributors.md:
```diff
@@ -209,24 +242,125 @@ getEntity is now a function that returns a Promise.
+There are two ways you can do this:
+
+1. You can run `yarn build:api-reports` from the root of the project and it will go through all of the existing API Reports and update them or create new ones as needed. This may take a while but is generally the best method if you are new to this.
+2. You can run `yarn build:api-reports plugins/<your-plugin-with-changes>` from the root and it will update the existing API Report or create a new one
```
```suggestion
2. You can run `yarn build:api-reports plugins/<your-plugin-with-changes>` from the root and it will update the existing API Report or create a new one.
```

awanlin replied:
Fixed 👍

adamdmharvey reviewed (commented):
👏🏻 nice improvement, @awanlin

jhaals reviewed (changes requested):
Really nice work with this @awanlin!

There are some really good improvements here but we currently have a large overlap with our CONTRIBUTORS.md file. I took this up with the rest of the maintainers and we suggest to instead update that file and empty out the contributors.md file on the docs site. It's a lot harder to find it on the docs page compared to the git repo where it's more likely to look for that information.

awanlin commented on CONTRIBUTING.md:
```diff
@@ -28,6 +100,163 @@ Also be sure to skim through our [ADRs](docs/architecture-decisions) to see if t
+
+In general, changesets are only needed for changes to packages within `packages/` or `plugins/` directories, and only for the packages that are not marked as `private`. Changesets are also not needed for changes that do not affect the published version of each package, for example changes to tests or in-line source code comments.
+
+Changesets **are** needed for new packages, as that is what triggers the package to be part of the next release. They are also needed for changes to `README.md` files so that the updates are reflected on the NPM page for the changed package.
```
@adamdmharvey, how does this wording look to your regarding needing changesets for `README.md` files?

awanlin commented:
Hi @jhaals, took some time to move everything into the `CONTRIBUTORS.md` file. I also deleted `docs/getting-started/contributors.md` and put a redirect to the `CONTRIBUTORS.md` file. 

When you have some time can you take another look, please?

awanlin commented:
Alright, seems we can't redirect using a URL. I've added back `docs/getting-started/contributors.md` and cleaned it up to act as a landing page with a link to the `CONTRIBUTORS.md` file. That should give us what we want. With this change in place the builds are finally green.

@jhaals, I promise this time things are ready for review 😉

jhaals reviewed (approved):
Thanks @awanlin! 👏

awanlin commented:
Awesom, really happy to have these changes in place now! Many thanks @jhaals and @adamdmharvey 🚀
